azurelinuxagent/common/cgroupapi.py

```diff
@@ -497,36 +497,32 @@ def create_cgroup(controller):
         self.track_cgroups(extension_cgroups)
 
         # Wait for process completion or timeout
-        timed_out, return_code = wait_for_process_completion_or_timeout(process, timeout)
-        process_output = read_output(stdout, stderr)
-
-        if return_code == 0:
-            # The process terminated in time and successfully
-            return extension_cgroups, process_output
-        else:
+        try:
+            process_output = handle_process_completion(process=process,
+                                                       command=command,
+                                                       timeout=timeout,
+                                                       stdout=stdout,
+                                                       stderr=stderr,
+                                                       error_code=error_code)
+        except ExtensionError as e:
             # The extension didn't terminate successfully. Determine whether it was due to systemd errors or
             # extension errors.
+            process_output = read_output(stdout, stderr)
             systemd_failure = self._is_systemd_failure(scope_name, process_output)
 
             if not systemd_failure:
-                # There was an extension error; it either timed out or returned a non-zero exit code.
-                if timed_out:
-                    raise ExtensionError("Timeout({0}): {1}\n{2}".format(timeout, command, process_output),
-                                         code=ExtensionErrorCodes.PluginHandlerScriptTimedout)
-                else:
-                    raise ExtensionError("Non-zero exit code: {0}, {1}\n{2}".format(return_code,
-                                                                                    command,
-                                                                                    process_output),
-                                         code=error_code)
+                # There was an extension error; it either timed out or returned a non-zero exit code. Re-raise the error
+                raise
             else:
                 # There was an issue with systemd-run. We need to log it and retry the extension without systemd.
+                exit_code = e.exit_code if isinstance(e, ExtensionOperationError) else e.code
                 add_event(AGENT_NAME,
                           version=CURRENT_VERSION,
                           op=WALAEventOperation.InvokeCommandUsingSystemd,
                           is_success=False,
                           message='Failed to run systemd-run for unit {0}.scope. '
                                   'Process exited with code {1} and output {2}'.format(scope_name,
-                                                                                       return_code,
+                                                                                       exit_code,
```

Comment:
but if the exception is not an ExtensionOperationError, then exit_code is not really an exit code... maybe you need to improve the message

Reply:
the idea would be to differentiate timeouts from actual errors

Reply:
Makes sense, updated the code